### PR TITLE
Support override of default value when using referenced schemas (#189)

### PR DESCRIPTION
**Pull request by martin-koch**
Solves #189 by copying the referenced schema and modifying the default value afterwards.

### CODE REVIEW

File: src/json-validator.cpp [hunk 1]
```diff
@@ -1276,6 +1281,17 @@ std::shared_ptr<schema> schema::make(json &schema,
 			// so this is the origial URI for this reference, the $ref-value has thus be resolved from it
 			auto id = uris.back().derive(attr.value().get<std::string>());
 			sch = root->get_or_create_ref(id);
+
+			const auto defaultAttr = schema.find("default");
```
Comment (pboettch) [inline src/json-validator.cpp:1285]: lower-case snake-case please
```suggestion
			const auto default_attr = schema.find("default");
```

File: src/json-validator.cpp [hunk 2]
```diff
@@ -630,6 +630,11 @@ class type_schema : public schema
 			sch.erase(attr);
 		}
 	}
+
+	void setDefaultValue(const json &defaultValue)
```
Comment (pboettch) [inline src/json-validator.cpp:634]:
```suggestion
	void set_default_value(const json &default_value)
```

File: test/issue-189-default-values.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+		// height must be 10 according to the default specified in the length definition while width must be 10 overridden by the width element
+		const auto expected = R"({"height":10,"width":20})"_json;
+		if (actual != expected) {
+			std::cerr << "Patch with defaults contains wrong value: '" << default_patch << "' instead of expected '" << expected.dump() << "'" << std::endl;
+			return 1;
+		}
+	}
+	return 0;
+}
```
Comment (pboettch) [inline test/issue-189-default-values.cpp:49]: Could you add a test where two stacked/recursive `$refs` are used? And see whether it still works. I'm afraid that there is some late evaluation of not-yet loaded references which will break this mechanism.

**Review by pboettch — CHANGES_REQUESTED** [review]

File: test/issue-189-default-values.cpp (hunk 3 above)
Comment (martin-koch) [reply test/issue-189-default-values.cpp:50]: Can you provide an example with stacked/recursive $refs and default values? I am not able to figure out a useful one...

Comment (pboettch) [reply test/issue-189-default-values.cpp:50]: In addition to that, it should use other files to see whether late evaluation works. (You might laugh, but I don't remember whether I do late evaluation of references or not...) Let me check your branch and see what I can do.

Comment (pboettch) [reply test/issue-189-default-values.cpp:50]: I adapted your example, it doesn't work if there is one more indirection. Could you check whether you can make it work. https://github.com/pboettch/json-schema-validator/tree/martin-koch/master

Comment (pboettch) [reply test/issue-189-default-values.cpp:50]: OK, I'm looking at it now, I think there is no need to copy the schema. We just need to get the earliest default value. I'll try.

Comment (pboettch) [reply test/issue-189-default-values.cpp:50]: Got it. I'll push on your master branch via this pull-request. Please test and review. And you're right, we need to copy the ref-schema when a (new) default is present.

Comment (pboettch) [reply test/issue-189-default-values.cpp:67]: Damn it. There is a corner case regarding the copy of the ref... When the same ref on a ref is used, the default value won't be used... hmm, I don't understand myself, let's think about it a bit more.

Comment (pboettch) [reply test/issue-189-default-values.cpp:70]: No, it's OK. The copy is actually what's need to be done. Thanks. If you're OK with these changes, could you update the README with this use-case and than I will happily merge it.

Comment (martin-koch) [reply test/issue-189-default-values.cpp:109]: I updated the readme and fixed an issue with the simpler example (segmentation fault) which occurred after change 40f5d29 since get_or_create_ref may return an instance of schema_ref **or** an instance of type_schema.

**Comment by martin-koch** [comment]
@pboettch are you okay with merging it?

**Comment by pboettch** [comment]
Merged. Thank you for your patience.

**Comment by pboettch** [comment]
All PRs have been closed due to me changing the main-branch to `main`. Please reissue your pull-request to that branch. Sorry and thanks.
